### DISCUSSION TITLE
Running test to validate user with token but test fails

### DISCUSSION THREAD

**Discussion by cansati** [pocketbase/pocketbase#998] [Q&A] [answered]
Hello everybody, I'm using (pocketbase@v0.8.0-rc2).

I want that only a user could see the content of "/api/hello-user" endpoint. I added the middleware `apis.RequireRecordAuth(),` while adding a Route. Afterwards I wrote the test codes below:
```
scenarios := []tests.ApiScenario{
{
			Name:            "try as authenticated user",
			Method:          http.MethodGet,
			Url:             "/api/hello-user",
			RequestHeaders:  map[string]string{"Authorization": "the.looooong.token"},
			ExpectedStatus:  200,
			ExpectedContent: []string{"Hello user"},
			TestAppFactory:  setupTestApp,
		},
```
But the logs of test was a fail:
```
api.go:99: [try as authenticated user] Expected status code 200, got 401
api.go:124: [try as authenticated user] Cannot find Hello user in response body {"code":401,"message":"The request requires valid record authorization token to be set.","data":{}}
```

I tried the Authorization token via Postman and It was working fine.
What am I missing?

**Comment by ganigeorgiev** [comment, marked as answer]
@cansati Are you sure that the token you are using is correct? I guess you are running against a test database, right? You can inspect the token payload in something like https://jwt.io/ and check if its claims are properly set and the record id exist. A valid authentication token claims should look something like this:

```json
{
  "id": "4q1xlclmfloku33",
  "type": "authRecord",
  "collectionId": "_pb_users_auth_",
  "exp": 2208985261
}
```

**Reply by ganigeorgiev** [reply]
btw. You can also check this comment in https://github.com/pocketbase/pocketbase/discussions/396#discussioncomment-3587783 for an example how to setup a test database (I'll make sure to document it properly with the final v0.8.0 docs).

**Reply by cansati** [reply]
@ganigeorgiev This was the token: eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJjb2xsZWN0aW9uSWQiOiJfcGJfdXNlcnNfYXV0aF8iLCJleHAiOjE2NjkzNzMzMjQsImlkIjoicWMxeHVnZmR3eGlkbWZyIiwidHlwZSI6ImF1dGhSZWNvcmQifQ.tIUJOHIJ5oFw9QoFNS8utzvgzWLR4cTATjuGTlZqC6w

"id" is matching with the user id inside the users collection. 

The payload seems fine to me but can you check too?.

The whole test code is:
```
// main_test.go
package main

import (
	"net/http"
	"testing"

	"fonksiyon-semsis.com/backend/hooks"
	"github.com/pocketbase/pocketbase/tests"
)

func TestHelloEndpoint(t *testing.T) {

	setupTestApp := func() (*tests.TestApp, error) {
		app, err := tests.NewTestApp()
		if err != nil {
			return nil, err
		}

		hooks.BindAppHooks(app, "./pb_public")

		return app, nil
	}

	scenarios := []tests.ApiScenario{
		{
			Name:            "try as authenticated user",
			Method:          http.MethodGet,
			Url:             "/api/hello-user",
			RequestHeaders:  map[string]string{"Authorization": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJjb2xsZWN0aW9uSWQiOiJfcGJfdXNlcnNfYXV0aF8iLCJleHAiOjE2NjkzNzMzMjQsImlkIjoicWMxeHVnZmR3eGlkbWZyIiwidHlwZSI6ImF1dGhSZWNvcmQifQ.tIUJOHIJ5oFw9QoFNS8utzvgzWLR4cTATjuGTlZqC6w"},
			ExpectedStatus:  200,
			ExpectedContent: []string{"Hello user"},
			TestAppFactory:  setupTestApp,
		},
	}

	for _, scenario := range scenarios {
		scenario.Test(t)
	}
}

```

**Reply by ganigeorgiev** [reply]
@cansati The token playload data is OK, but looking at your code sample it seems that you are using the default test database, aka - https://github.com/pocketbase/pocketbase/tree/master/tests/data and the auth record associated with the token you are testing doesn't exist there.

My guess is that you are using an auth token from your actual application data, right? If that's the case, then it won't work and the recommended way is to have a fixed test `pb_data` that will act as "fixtures" for your tests (aka. you'll have test users and test collection that you can use safely in your tests). To do that you need to specify a `test_pb_data` directory when calling `tests.NewTestApp("./test_pb_data")` (see the example in https://github.com/pocketbase/pocketbase/discussions/396#discussioncomment-3587783).

Or to summarize:

1. Create your `test_pb_data`, either by copying your existing `pb_data` OR creating a completely new test data with the help of the Admin UI:
    ```sh
    ./go run main.go serve --dir="./test_pb_data/"
    # open 127.0.0.1:8090 and setup your test data (test users and test collection records)
    ```

2. Change your code to load the `test_pb_data` directory:
    ```go
    // main_test.go
    package main

    import (
        "net/http"
        "testing"

        "fonksiyon-semsis.com/backend/hooks"
        "github.com/pocketbase/pocketbase/tests"
    )

    func TestHelloEndpoint(t *testing.T) {

        setupTestApp := func() (*tests.TestApp, error) {
            app, err := tests.NewTestApp("./test_pb_data")
            if err != nil {
                return nil, err
            }

            hooks.BindAppHooks(app, "./pb_public")

            return app, nil
        }

        scenarios := []tests.ApiScenario{
            {
                Name:            "try as authenticated user",
                Method:          http.MethodGet,
                Url:             "/api/hello-user",
                // !!! change the token with one for your test users (I'll add another comment with guide how to do that)
                RequestHeaders:  map[string]string{"Authorization": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJjb2xsZWN0aW9uSWQiOiJfcGJfdXNlcnNfYXV0aF8iLCJleHAiOjE2NjkzNzMzMjQsImlkIjoicWMxeHVnZmR3eGlkbWZyIiwidHlwZSI6ImF1dGhSZWNvcmQifQ.tIUJOHIJ5oFw9QoFNS8utzvgzWLR4cTATjuGTlZqC6w"},
                ExpectedStatus:  200,
                ExpectedContent: []string{"Hello user"},
                TestAppFactory:  setupTestApp,
            },
        }

        for _, scenario := range scenarios {
            scenario.Test(t)
        }
    }
    ```

**Reply by cansati** [reply]
Worked successfully. Thank you so much @ganigeorgiev

**Reply by ganigeorgiev** [reply]
For how to generate tokens for your test users - the best way would be to create a token with `exp` claim that is far in the future (eg. 2030). To do that this currently requires a little manual work (I have added it in my todo to add a test helper for this):

1. Find the `tokenKey` of your test user record. You can use the sqlite3 cli tool for that:
    ```sh
    sqlite3 test_pb_data/data.db "select tokenKey from users where id = 'YOUR_TEST_USER_ID'"
    ```

2. Find the shared `recordAuthToken` secret from the test app settings. Again, you can use the sqlite3 cli tool:
    ```sh
    sqlite3 test_pb_data/data.db "select json_extract(value, '$.recordAuthToken.secret') from _params where key='settings'";
    ```

3. Go to https://jwt.io/ and use your previous token for scaffolding.

4. Adjust the "id" part to match with your test record id and change the "exp" claim to be 2208985261 (eg. 2040-01-01 01:01:01).

5. In the "VERIFY SIGNATURE" field place first the user tokenKey from step 1) and right after it concatenate the settings secret from step 2)

6. Use the generated token in your tests. This token could be used for a valid auth checks. For expired auth token checks you can generate similar token but with "exp" claim with some date in the past.
